Test : 좌석 예약 동시성 테스트 케이스 추가

veniharuka opened the pull request:
### PR 타입(하나 이상의 PR 타입을 선택해주세요)
* 기능 추가 상세 설명 필요시 작성하기
  - [ ] 기능 추가
  - [ ] 기능 수정
  - [ ] 기능 삭제
  - [ ] 버그 수정
  - [x] 테스트 추가/수정
  - [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트

### 반영 브랜치
ex) test/ticketing/pre-concurrency/cmj-> main

### 변경 사항
- 동일 좌석에 대한 동시 예매 테스트
  여러 사용자가 동시에 같은 좌석을 예매할 경우, 단 1건의 성공만 허용하는지 확인.
  동시성 문제를 방지하기 위해 데이터베이스 수준에서의 락이나 @Transactional 같은 트랜잭션 관리가 필요.

- 서로 다른 좌석에 대한 동시 예매 테스트
  50명이 서로 다른 좌석을 동시에 예매할 때 모든 요청이 성공해야 함.
  동시성 처리가 좌석별로 올바르게 동작하는지 검증.

- 동시성 제어 없이 동일 좌석 동시 예매
  외래 키 제약조건 및 유니크 키를 제거한 상태에서 동일 좌석 예매 시 중복이 발생하는지 확인.
  동시성 제어의 필요성을 입증하기 위한 테스트.

- 시스템 장애 시나리오
  좌석 상태는 갱신되었지만, 티켓 생성이 실패한 경우를 시뮬레이션.
  장애 상황에서 데이터 정합성을 유지하는 방법 검증.

- 이미 예매된 좌석에 대한 중복 예매
  기존에 예약된 좌석에 대해 또 다른 사용자가 예매를 시도하는 경우 실패해야 함.

- 좌석 상태와 예매 정보의 불일치 검증
  좌석이 예약된 상태지만, 예매 정보는 존재하지 않는 비정상 상태를 테스트.
  데이터 정합성을 위한 검증 로직의 필요성을 강조.

- 결제 성공 후 예매 저장 실패
  결제는 성공했으나 예매 정보 저장이 실패한 경우 데이터 일관성을 보장하는지 확인.
  트랜잭션 롤백이나 보상 트랜잭션 설계 필요.

### 검증 여부
- [ ] postman으로 api 테스트 완료했습니다.
- [ ] 로컬 test 환경에서 JUnit 테스트 코드 실행 및 통과.


50명이 각기 다른 좌석에 예매 했을 때 결과
===== 예매 테스트 결과 분석 =====
```
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 1. 전체 요약
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - - 총 소요 시간: 1049.1756ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - - 전체 시도: 50 건
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - - 성공: 50 건
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - - 실패: 0 건
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - - 평균 처리 시간: 792.988052ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 
2. 예매 성공 내역
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test34@test.com, 좌석: 34, 처리시간: 566.5104ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test33@test.com, 좌석: 33, 처리시간: 567.6093ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test15@test.com, 좌석: 15, 처리시간: 568.0817ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test27@test.com, 좌석: 27, 처리시간: 568.5411ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test13@test.com, 좌석: 13, 처리시간: 568.9604ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test48@test.com, 좌석: 48, 처리시간: 608.3611ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test29@test.com, 좌석: 29, 처리시간: 622.3666ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test35@test.com, 좌석: 35, 처리시간: 629.9786ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test24@test.com, 좌석: 24, 처리시간: 635.9743ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test5@test.com, 좌석: 5, 처리시간: 638.6897ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test21@test.com, 좌석: 21, 처리시간: 647.4151ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test6@test.com, 좌석: 6, 처리시간: 651.5145ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test20@test.com, 좌석: 20, 처리시간: 663.5455ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test47@test.com, 좌석: 47, 처리시간: 669.7208ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test1@test.com, 좌석: 1, 처리시간: 677.7863ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test41@test.com, 좌석: 41, 처리시간: 690.7065ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test10@test.com, 좌석: 10, 처리시간: 691.4166ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test42@test.com, 좌석: 42, 처리시간: 693.2419ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test30@test.com, 좌석: 30, 처리시간: 712.7781ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test22@test.com, 좌석: 22, 처리시간: 716.1216ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test12@test.com, 좌석: 12, 처리시간: 732.2995ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test3@test.com, 좌석: 3, 처리시간: 742.1748ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test2@test.com, 좌석: 2, 처리시간: 746.2451ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test7@test.com, 좌석: 7, 처리시간: 757.1002ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test11@test.com, 좌석: 11, 처리시간: 768.1895ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test37@test.com, 좌석: 37, 처리시간: 780.5776ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test32@test.com, 좌석: 32, 처리시간: 782.1386ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test9@test.com, 좌석: 9, 처리시간: 796.7142ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test40@test.com, 좌석: 40, 처리시간: 804.8938ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test39@test.com, 좌석: 39, 처리시간: 822.2813ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test17@test.com, 좌석: 17, 처리시간: 828.3368ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test28@test.com, 좌석: 28, 처리시간: 832.4736ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test45@test.com, 좌석: 45, 처리시간: 846.1321ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test18@test.com, 좌석: 18, 처리시간: 869.347ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test23@test.com, 좌석: 23, 처리시간: 879.3092ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test16@test.com, 좌석: 16, 처리시간: 895.9736ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test19@test.com, 좌석: 19, 처리시간: 903.5667ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test38@test.com, 좌석: 38, 처리시간: 911.2318ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test8@test.com, 좌석: 8, 처리시간: 943.6347ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test44@test.com, 좌석: 44, 처리시간: 967.3408ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test14@test.com, 좌석: 14, 처리시간: 999.5148ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test50@test.com, 좌석: 50, 처리시간: 1000.6825ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test25@test.com, 좌석: 25, 처리시간: 1010.1774ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test4@test.com, 좌석: 4, 처리시간: 1020.1685ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test46@test.com, 좌석: 46, 처리시간: 1020.7027ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test43@test.com, 좌석: 43, 처리시간: 1029.1648ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test36@test.com, 좌석: 36, 처리시간: 1037.82ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test31@test.com, 좌석: 31, 처리시간: 1039.9129ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test49@test.com, 좌석: 49, 처리시간: 1044.0478ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test26@test.com, 좌석: 26, 처리시간: 1047.9302ms
```

veniharuka committed:
Test : 좌석 예약 동시성 테스트 케이스 추가

-동일 좌석에 대한 동시 예매 테스트
-서로 다른 좌석에 대한 동시 예매 테스트
-동시성 제어 없이 동일 좌석 동시 예매(수정 중)
-시스템 장애 시나리오
-이미 예매된 좌석에 대한 중복 예매
-좌석 상태와 예매 정보의 불일치 검증
-결제 성공 후 예매 저장 실패
-티켓 예외 처리 위한 TicketException 클래스 추가

gentle-tiger commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
-      Long.class
-    );
-    assertEquals(2, memberCount, "Both members should be inserted successfully.");
+  @DisplayName("50명이 동시에 같은 좌석을 예매할 경우, 단 1건만 성공해야 한다")
```
일반적인 동시성 전 문제파악 테스트 코드로 추가하면 좋다고 생각함~ 👍

gentle-tiger commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+  }
+
+  @Test
+  @DisplayName("좌석 상태와 예매 정보 불일치 검증")
```
false로 설정된 좌석에 대해서도 검증하는 테스트 코드도 좋다고 생각함~👍

chuseok replied:
TicketServiceTest에 옮기는게 좋을 듯! b

veniharuka replied:
완료하였습니다

veniharuka replied:
반영하였습니다.

gentle-tiger commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+  }
+
+  @Test
+  @DisplayName("결제 성공 후 예매 정보 저장 실패 시나리오")
```
이 시나리오로 좋다고 생각함 ~ 👍

gentle-tiger commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+  }
+
+  @Test
+  @DisplayName("50명이 동시에 서로 다른 좌석을 예매할 경우, 모든 예매가 성공해야 한다")
```
모든 예매가 정상적으로 처리되는지 확인하는 테스트 코드도 좋다고 생각함~ 👍

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+import org.springframework.transaction.annotation.Transactional;
 
 @ExtendWith(SpringExtension.class)
-@ContextConfiguration(classes = {
```
이거 없으면 밑에 44, 47라인 @Autowired 빨간 줄 뜨면서 사용이 안될텐데 어떻게 했어?

veniharuka replied:
37라인 보면 @ContextConfiguration(classes = {AppConfig.class})이 있습니다.
수정하고 추가하는 과정에서 깃에서 잡힌 거 같습니다.

chuseok replied:
CTRL + ALT + L 누르면 자동 정렬되는데 구글 컨벤션 적용해야 할듯!

veniharuka replied:
구글 컨벤션 설정을 잊고있긴했습니다! 적용토록 하겠습니다.

veniharuka replied:
![image](https://github.com/user-attachments/assets/e82e06b9-6ca2-4bf9-ba08-e5651ca937e5)
맞게 되었는지 확인 부탁드립니다.

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
-        +
-        "(1, 'test1@gmail.com', true, 'test1', 'password1', '01011111111', '111111', 'USER')");
+  void setUp() {
+    cleanUpAll();
```
이건 AfterEach에서 초기화를 한번 하니까 제거해도 문제없이 동작할 거 같은데 왜 추가했는지 궁금해

veniharuka replied:
AfterEach에서  초기화를 하면 Mysql에서 관련 데이터를 확인할 수 없어서 AfterEach의 cleanUpAll을 주석처리하고 시작할 때 초기화 진행하고 다른 테스트를 진행하기 위해 있습니다.  AfterEach와 setup의  cleanUpAll은 주석처리/해제를 진행하며 하고자하는 테스트에 맞춰 사용하느라 있었는데 AfterEach의 cleanUpAll을 차라리 없애는 게 낫다는 생각이 드네요

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+    cleanUpAll();
+  }
 
-    // musical 데이터 삽입
```
데이터 삽입하는 부분은 건들 필요가 없어서 원래대로 살리는게 좋을 거 같아

veniharuka replied:
1.코드 가독성 향상
2.관심사 분리
- 데이터 초기화 로직을 별도의 메서드로 분리
- 테스트 설정(setUp)과 데이터 생성 로직 분리
- 코드의 모듈성과 유지보수성 증가
3.재사용성
- 동일한 테스트 데이터 초기화 로직을 여러 테스트 메서드에서 활용 가능
- 테스트 간 일관된 초기 상태 보장

위와 같은 이유로 initializeTestData 메서드를 만들고 사용한 것이었습니다. 
그럼에도 원래대로 되돌리는 게 좋은 코드인가요? 잘 몰라서 여쭙니다.

gentle-tiger replied:
가독성과 관심사 분리가 중요하지만, 협업을 위해 기존 틀을 유지하는 것이 더 적합할 것 같습니다. 코드가 길지 않으니 이 정도는 괜찮다고 생각됩니다..!

chuseok replied:
분리하는게 나쁘다는게 아니고 

```
jdbcTemplate.execute(
      "insert into musical (id, place, place_detail, ranking, running_time, ticketing_end_date, ticketing_start_date, title, detail_image_url, notice_image_url, place_image_url, poster_image_url) values "
        +
        "(1, '서울', '서울 아트센터', 1, 150, '2024-10-16', '2024-10-01', '킹키부츠', null, null, null, null)");
```

이거를 

```
jdbcTemplate.execute(
      "INSERT INTO musical (id, place, place_detail, ranking, running_time, ticketing_end_date, "
            +
            "ticketing_start_date, title) VALUES " +
            "(1, '서울', '서울 아트센터', 1, 150, '2024-10-16', '2024-10-01', '킹키부츠')"
    );
```

이렇게 대문자로 바꿀 필요가 없다는 뜻이야 이렇게 되면 리뷰할 때 보기 너무 어려워

`// section 데이터 삽입`
이거를
`// 섹션 데이터 삽입`
이런 식으로 바꾸는 것들도

veniharuka replied:
그 전의 형태로 수정하였는데, 커밋 후 확인 부탁드립니다.

veniharuka replied:
소문자로 수정하였습니다

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
-  @AfterEach
-  void tearDown() {
-    cleanUpAll();
+    // 50개의 좌석 데이터 삽입
```
50개의 다른 좌석에 50명이 예매하는 경우 테스트하는 코드랑 같이 TicketServiceTest에 옮기는게 좋을 듯!

veniharuka replied:
완료하였습니다

chuseok replied:
여기 있는 코드 지워주세요

veniharuka replied:
삭제하였습니다

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+  }
+
+  @Test
+  @DisplayName("50명이 동시에 서로 다른 좌석을 예매할 경우, 모든 예매가 성공해야 한다")
```
위에서 말한대로 TicketServiceTest에 옮기는게 좋을 듯! b

veniharuka replied:
완료하였습니다

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+    long testStartTime = System.nanoTime();
+    startLatch.countDown();
+
+    boolean completed = doneLatch.await(30, TimeUnit.SECONDS);
```
30으로 설정한 이유가 있을까? 이유가 있어서 사용한건지 궁금해

veniharuka replied:
GPT가 설정해준 값이긴 합니다.
await의 구체적인 시간을 두는 이유에 대해서는 아래와 같습니다.
1. 테스트의 최대 대기 시간 제한
무한정 대기하는 것을 방지
데드락이나 예상치 못한 문제 발생 시 테스트가 무한정 멈추는 것을 막기 위함
2.실제 동시성 테스트의 합리적인 타임아웃
대부분의 동시성 작업은 몇 초 이내에 완료되어야 함
30초는 충분히 여유로운 시간

-다른 값을 넣을 수 있지만, 고려해야 할 점들:
   너무 짧으면 (예: 1초) 테스트가 불완전하게 종료될 수 있음
   너무 길면 (예: 5분) 실제 문제 발생 시 오랫동안 대기하게 됨

*일반적으로 추천되는 범위:
스레드 수와 작업 복잡도에 따라 10~60초 사이
대부분의 동시성 테스트는 30초면 충분

라고 답변을 하는데
https://herojoon-dev.tistory.com/216 해당 글을 참고하면 30초도 긴 것 같습니다 5~10초면 될 거 같아보여요.

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+  void testConcurrentDifferentSeatBooking() throws InterruptedException {
+    int numberOfThreads = 50;
+    ExecutorService executor = Executors.newFixedThreadPool(numberOfThreads);
+    CountDownLatch readyLatch = new CountDownLatch(numberOfThreads);
```
구분해서 사용한 이유가 있어? 꼭 필요한 이유가 있던게 아니면 구분하지 않는게 좋은 거 같아

chuseok replied:
testConcurrentSameSeatBooking() 여기서도 사용했는데 정리해서 올려주면 좋을 듯!

veniharuka replied:
단일 래치 사용이 적합한 상황으로는:
1. 모든 스레드가 동일한 작업을 수행할 때
2. 작업 시작 시점의 세밀한 동기화가 필요하지 않을 때
3. 간단한 동시성 테스트

통합된 단일 래치는 다음과 같은 문제가 발생할 수 있습니다.
1. 스레드 준비 시점의 불일치
2. 작업 시작의 비동기성
3. 예외 처리 시 정확한 동기화 어려움

동시성 테스트에서는 각 단계를 명확히 구분하는 다중 래치 사용하면 얻는 이점으로는
1. 스레드 준비 상태 동기화
2. 정확한 동시 시작 보장
3. 작업 완료 추적의 명확성
https://devidea.tistory.com/89  와 https://goalinnext.tistory.com/107 보면 다중/단일 래치를 사용하고 있는데, 어느 상황에 맞춰 사용하고 있는지를 알기 위해 다중 래치가 낫지 않나라는 개인적인 생각입니다.

chuseok replied:
done이나 ready는 왜 사용하는지 알겠는데 `startLatch.countDown();` 은 다른 Latch와 다르게 반대로 쓰고 있는데 이렇게 한 의도를 잘 모르겠어

veniharuka replied:
**readyLatch (10 → 0)**
- 모든 스레드가 실행 준비가 완료되었음을 알리는 용도
- 각 스레드가 준비되면 countDown()을 호출
- 메인 스레드는 await()로 모든 스레드가 준비될 때까지 대기

**doneLatch (10 → 0)**
- 모든 스레드의 작업 완료를 확인하는 용도
- 각 스레드가 작업을 마치면 countDown() 호출
- 메인 스레드는 await()로 모든 작업이 끝날 때까지 대기

**startLatch (1 → 0) - 여기가 다른 부분**
- 시작 신호로 사용됨
- 초기값이 1이고, 메인 스레드가 단 한 번만 countDown() 호출
- 모든 작업 스레드들은 await()로 이 신호를 기다림

이렇게 함으로써 모든 스레드가 정확히 같은 시점에 시작하도록 동기화
startLatch를 다른 Latch들처럼 10에서 시작하게 되면, 각 스레드가 개별적으로 시작하게 되어 진정한 의미의 동시성 테스트가 어려워집니다. 1에서 시작하고 메인 스레드가 한 번에 신호를 주는 방식을 통해 더 정확한 동시성 테스트가 가능해집니다.

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+    CountDownLatch startLatch = new CountDownLatch(1);
+    CountDownLatch doneLatch = new CountDownLatch(numberOfThreads);
+
+    ConcurrentHashMap<String, BookingResult> results = new ConcurrentHashMap<>();
```
ConcurrentHashMap 이런 것도 제대로 알고 썼을까?

veniharuka replied:
GPT를 사용하여 만들어서 처음보는 자료구조였습니다. 동시성 ConcurrentHashMap 로 구글링 하였을 때,
ConcurrentHashMap 사용을 권장하는 경우는
1. 다중 스레드 환경
2. 동시 데이터 수정 필요할 때
3. 읽기 작업이 쓰기 작업보다 많은 경우
ConcurrentHashMap 과 HashMap 의 차이점은 아래와 같다는 것을 알았습니다.
- 1) Thread Safe
주요 차이점은 ConcurrentHashMap는 내부적 동기화 때문에 스레드가 Safe합니다. HashMap는 내부적으로 동기화되지 않고 스레드로부터 안전하지 않습니다. HashMap 메서드를 사용하여 외부에서 동기화 할 수 있습니다.
- 2) Internal Structure(내부구조)
ConcurrentHashMap의 모든 작업이 동기화되는 것은 아닙니다. 추가 및 삭제와 같은 수정 작업만 동기화됩니다. 읽기 작업은 동기화되지 않습니다. 이렇게 하면 ConcurrentHashMap이 외부에서 동기화된 HashMap보다 동시 다중 스레드 응용 프로그램에 대한 첫 번째 선택 맵이 됩니다.
- 3) Null Keys And Null Values
HashMap은 최대 하나의 null 키와 임의의 수의 null 값을 허용합니다.ConcurrentHashMap은 null 키와 null 값도 허용하지 않습니다.
- 4) Fail-Fast Vs Fail-Safe
HashMap에 의해 반환된 반복자는 본질적으로 빠른 속도입니다. 반복자 생성 후 맵이 수정되면 ConcurrentModificationException이 발생하기 때문입니다. ConcurrentHashMap에 의해 반환된 반복자는 본질적으로 안전합니다. iterator 생성 후 맵이 수정되면 예외가 발생하지 않습니다.
- 5) Performance(성능)
ConcurrentHashMap에 대한 수정 작업만 동기화됩니다. 따라서 ConcurrentHashMap에 대한 추가 또는 제거 작업은 HashMap보다 느립니다. ConcurrentHashMap 및 HashMap 모두에 대한 읽기 작업은 두 맵의 읽기 작업이 동일한 성능을 제공합니다.
https://applepick.tistory.com/124 -> 리뷰에 참고한 글
https://wildeveloperetrain.tistory.com/271 -> 이 블로그가 더 좋아보입니다.

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+    executor.shutdown();
+
+    log.info("테스트 결과 분석 시작");
+    analyzeResults(results, bookingOrder, testEndTime - testStartTime);
```
어떤 결과값이 나오는지 궁금한데 pr 템플릿에 작성 좀 부탁해

veniharuka replied:
===== 예매 테스트 결과 분석 =====
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 1. 전체 요약
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - - 총 소요 시간: 1049.1756ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - - 전체 시도: 50 건
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - - 성공: 50 건
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - - 실패: 0 건
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - - 평균 처리 시간: 792.988052ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 
2. 예매 성공 내역
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test34@test.com, 좌석: 34, 처리시간: 566.5104ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test33@test.com, 좌석: 33, 처리시간: 567.6093ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test15@test.com, 좌석: 15, 처리시간: 568.0817ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test27@test.com, 좌석: 27, 처리시간: 568.5411ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test13@test.com, 좌석: 13, 처리시간: 568.9604ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test48@test.com, 좌석: 48, 처리시간: 608.3611ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test29@test.com, 좌석: 29, 처리시간: 622.3666ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test35@test.com, 좌석: 35, 처리시간: 629.9786ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test24@test.com, 좌석: 24, 처리시간: 635.9743ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test5@test.com, 좌석: 5, 처리시간: 638.6897ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test21@test.com, 좌석: 21, 처리시간: 647.4151ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test6@test.com, 좌석: 6, 처리시간: 651.5145ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test20@test.com, 좌석: 20, 처리시간: 663.5455ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test47@test.com, 좌석: 47, 처리시간: 669.7208ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test1@test.com, 좌석: 1, 처리시간: 677.7863ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test41@test.com, 좌석: 41, 처리시간: 690.7065ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test10@test.com, 좌석: 10, 처리시간: 691.4166ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test42@test.com, 좌석: 42, 처리시간: 693.2419ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test30@test.com, 좌석: 30, 처리시간: 712.7781ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test22@test.com, 좌석: 22, 처리시간: 716.1216ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test12@test.com, 좌석: 12, 처리시간: 732.2995ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test3@test.com, 좌석: 3, 처리시간: 742.1748ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test2@test.com, 좌석: 2, 처리시간: 746.2451ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test7@test.com, 좌석: 7, 처리시간: 757.1002ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test11@test.com, 좌석: 11, 처리시간: 768.1895ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test37@test.com, 좌석: 37, 처리시간: 780.5776ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test32@test.com, 좌석: 32, 처리시간: 782.1386ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test9@test.com, 좌석: 9, 처리시간: 796.7142ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test40@test.com, 좌석: 40, 처리시간: 804.8938ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test39@test.com, 좌석: 39, 처리시간: 822.2813ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test17@test.com, 좌석: 17, 처리시간: 828.3368ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test28@test.com, 좌석: 28, 처리시간: 832.4736ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test45@test.com, 좌석: 45, 처리시간: 846.1321ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test18@test.com, 좌석: 18, 처리시간: 869.347ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test23@test.com, 좌석: 23, 처리시간: 879.3092ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test16@test.com, 좌석: 16, 처리시간: 895.9736ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test19@test.com, 좌석: 19, 처리시간: 903.5667ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test38@test.com, 좌석: 38, 처리시간: 911.2318ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test8@test.com, 좌석: 8, 처리시간: 943.6347ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test44@test.com, 좌석: 44, 처리시간: 967.3408ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test14@test.com, 좌석: 14, 처리시간: 999.5148ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test50@test.com, 좌석: 50, 처리시간: 1000.6825ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test25@test.com, 좌석: 25, 처리시간: 1010.1774ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test4@test.com, 좌석: 4, 처리시간: 1020.1685ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test46@test.com, 좌석: 46, 처리시간: 1020.7027ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test43@test.com, 좌석: 43, 처리시간: 1029.1648ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test36@test.com, 좌석: 36, 처리시간: 1037.82ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test31@test.com, 좌석: 31, 처리시간: 1039.9129ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test49@test.com, 좌석: 49, 처리시간: 1044.0478ms
2025-01-15 11:43:07 [Test worker] INFO  c.k.w.t.s.TicketServiceConcurrencyTest - 사용자: test26@test.com, 좌석: 26, 처리시간: 1047.9302ms

chuseok replied:
pr 템플릿에 부탁해!

veniharuka replied:
PR 템플릿에 작성하였습니다

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+      String.format("동시성 제어 부재로 인해 중복 예매(%d건)가 발생해야 함", ticketCount));
+  }
+  @Test
+  @DisplayName("이미 예매된 좌석에 대한 중복 예매 시도")
```
이런 건 TicketServiceTest에 실패하는 경우로 이미 작성돼 있을 듯!

veniharuka replied:
해당 테스트 메서드 삭제하도록 하겠습니다

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+
+  @Test
+  @DisplayName("좌석 상태와 예매 정보 불일치 검증")
+  void testSeatStatusAndTicketMismatch() {
```
TicketServiceTest에 옮기는게 좋을 듯! b

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+  }
+
+  @Test
+  @DisplayName("결제 성공 후 예매 정보 저장 실패 시나리오")
```
어떤 부분에서 결제가 성공된 시나리오인지 모르겠는데 어디서 하고 있는거야?

veniharuka replied:
고객이 특정 좌석에 대해 결제를 성공했는데, 그와 관련된 좌석 및 티켓 정보가 DB에 저장되지 않은 경우가 있지 않을까 싶어 만든 테스트 케이스인데 시나리오를 다시 검토하고 수정해야할 것으로 사료됩니다.

chuseok replied:
지워주세요~

veniharuka replied:
삭제한 줄 알았네요. 감사합니다

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
   }
 
   @Test
-  @DisplayName("단건 티켓을 예매할 경우, 예매가 성공한다.")
```
왜 삭제했는지 궁금해

veniharuka replied:
개인적으로 생각하는 동시성 시나리오를 진행하고자 처음에 전부 삭제했습니다.
별다른 이유는 없는데 필요한 경우라면 살리겠습니다 :)

gentle-tiger replied:
한 명의 사용자가 하나의 좌석을 예매하는 테스트 케이스를 통해, **기본적인 예매 기능의 정상 동작 여부를 검증**할 수 있는 사례로 판단되어, 이를 유지하는 것이 바람직하다고 생각합니다.!

chuseok replied:
이거 살려주세요!

veniharuka replied:
살렸습니다! TicketServiceTest로 옮기면 되나요?

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
-    Long ticketCount = jdbcTemplate.queryForObject(
-      "select count(*) from ticket where seat_id = 1 and member_id = 1",
-      Long.class
+  @DisplayName("결제 완료 후 동시 티켓 할당으로 인한 중복 예매 발생 시나리오")
```
어디서 동시에 티켓을 할당하는지 보이지 않는데 어디서 하고 있어?

chuseok replied:
이런 상황을 테스트 하는건 좋다고 생각해b

veniharuka replied:
동시" 티켓 할당하는 시나리오를 구상했고, 그렇게 짰다고 생각했지만, 실제로는 단일 스레드에서 순차적으로 실행하는 테스트 케이스라는 것을 알았습니다. 하여 코드를 수정하였고 아래와 같습니다.

// 동시성 제어를 위한 부분
CountDownLatch startLatch = new CountDownLatch(1);
CountDownLatch doneLatch = new CountDownLatch(2);

ExecutorService executorService = Executors.newFixedThreadPool(2);

// 두 개의 스레드(태스크)를 동시에 실행
executorService.submit(firstUserBookingTask);
executorService.submit(secondUserBookingTask);

// 동시 시작을 위한 메커니즘
startLatch.countDown(); // 이 순간 두 스레드가 동시에 티켓 예매 시도

1. ExecutorService로 두 개의 스레드 생성
2. startLatch를 통해 정확히 동시에 시작
3. 각 스레드에서 ticketService.bookTicket() 메서드 동시 호출

이 방식으로 두 사용자가 거의 동시에 같은 좌석에 대해 티켓 예매를 시도하게 됩니다.
핵심은 startLatch.countDown()이 호출되는 순간 두 스레드가 거의 동시에 티켓 예매 로직을 실행한다는 점입니다.

chuseok replied:
잘 고친듯 b 그래서 테스트 결과로 중복 예매가 발생했을까?

veniharuka replied:
데이터베이스의 유니크 제약조건으로 하나의 좌석에 중복 예매는 실패했습니다.
test1@test.com이 먼저 결제를 완료 후 test2@test.com이 나중에 결제를 완료 하였음에도 티켓 발급 시도에서는 test2가 먼저 성공하고 test1은 실패하는 결과를 확인함으로 동시성 제어가 필요하다는 것을 확인할 수 있었습니다.

chuseok replied:
이거 결제 직후 티켓 동시예매가 아니라 동일한 좌석에 여러명이 예매할 때랑 상황이 같아

시나리오가 그냥 동시 예매 상황인 거 같은데 이거 테스트코드 지우고, 우리 동일한 좌석에 여러 사용자가 예매할 때 시나리오의 테스트 코드에 스레드 동시실행 부분만 추가해서 수정하면 될 거 같아

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+    log.info("- 첫 번째 결제 사용자: {}", firstUser);
+    log.info("- 두 번째 결제 사용자: {}", secondUser);
+
+    // 1. 첫 번째 사용자 결제 시뮬레이션
```
1. firstUserId를 정의만 해도 될 거 같은데 왜 조회를 해오는지 궁금하고
2. 단순 조회해서 정의하건데 주석의 의도가 첫번째 사용자가 결제를 하는 작업인 것처럼 보여서 좀 헷갈리는 거 같아
3. 그리고 전체적으로 봤을 때, log에 표시된 내용인데 중복해서 주석으로 처리하는 부분이 불필요한 작업인 거 같아

veniharuka replied:
리뷰해주신 testDoubleBookingAfterPayment 메서드를 수정하게되었습니다.
재커밋했을 때 의문인 점 리뷰 달아주시면 감사하겠습니다.

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
-  }
 
-  @Test
-  @DisplayName("2명의 사용자가 동시에 티켓을 예매할 경우, 단 1건의 예매만 성공한다.")
```
왜 기존에 있는 코드를 삭제하고 testConcurrentSameSeatBookingWithoutLock() 를 만들려고 했어?

veniharuka replied:
이 또한 기존에 있던 동시성 테스트 코드가 다소 잘못되었다고 멋대로 생각해서
개인적으로 생각하는 동시성 시나리오를 진행하고자 처음에 전부 삭제했습니다.
별다른 생각없이 진행한 점은 잘못됐다고 생각되네요

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+      String.format("Seat %d should not be available after booking", seatId));
+  }
+
+  private static class BookingResult {
```
코드가 길어지면 잘 보이지 않아서 service 패키지 밑에 새로운 java를 생성하는게 좋은 거 같아!

chuseok replied:
테스트 결과를 구분해서 사용한거 좋다 b

veniharuka replied:
Ticket - dto - BookingResult.java 로 생성하였습니다. 감사합니다

veniharuka committed:
Refactor(dto): 코드 가독성을 위해 BookingResult 클래스를 분리

-BookingResult 클래스를 TicketServiceConcurrencyTest에서 dto 패키지로 이동.
-DTO 관련 로직을 별도의 패키지로 분리하여 코드 구조를 개선.

veniharuka committed:
Refactor(test): testConcurrentDifferentSeatBooking,testConcurrentSameSeatBooking 테스트 메서드를 TicketServiceTest로 이동 및 설정 수정

- testConcurrentDifferentSeatBooking과 testConcurrentSameSeatBooking 메서드를 TicketServiceConcurrencyTest.java에서 TicketServiceTest.java로 이동.

veniharuka committed:
Refactor(test): testSeatStatusAndTicketMismatch 테스트 메서드의 클래스 이동

-testSeatStatusAndTicketMismatch 테스트 메서드를TicketServiceConcurrencyTest.java에서 TicketServiceTest.java로 이동.

veniharuka committed:
Refactor(test): TicketServiceConcurrencyTest 클래스의 테스트 재구조화, 불필요한 메서드 제거 및 코드 가독성 개선

1.다음 테스트 메서드를 TicketServiceConcurrencyTest.java에서 TicketServiceTest.java로 이동:
-testSeatStatusAndTicketMismatch
-testConcurrentDifferentSeatBooking
-testConcurrentSameSeatBooking
2.테스트를 기능에 따라 올바르게 분류하여 재구조화 진행.
3.불필요한 테스트 메서드 제거:
-testConcurrentSameSeatBookingWithoutLock
-testSeatStatusAndTicketMismatch : 해당 테스트는 중복되거나 관련성이 낮아 제거.
4. testDoubleBookingAfterPayment 테스트 메서드 수정으로 로직과 신뢰성을 개선.
5.BookingResult 클래스를 ticket-dto-BookingResult.java로 분리하여 유지보수성과 관심사 분리 원칙을 준수.

chuseok commented on src/main/java/com/kb/wallet/ticket/exception/TicketException.java:
```diff
@@ -0,0 +1,27 @@
+package com.kb.wallet.ticket.exception;
+
+public class TicketException {
```
exception 으로는 global에서 따로 관리해서 이렇게 쓰지말고 globalExceptionHandler CustomException으로 관리해야돼

veniharuka replied:
반영하였습니다.

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceTest.java:
```diff
@@ -61,6 +93,7 @@ class TicketServiceTest {
   void setUp() {
     MockitoAnnotations.openMocks(this);
 
+    // 목 객체 설정
```
불필요한 주석이라 제거하면 좋겠습니다

veniharuka replied:
반영하였습니다.

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceTest.java:
```diff
@@ -73,6 +106,62 @@ void setUp() {
     when(seat2.getSchedule()).thenReturn(schedule);
     when(schedule.getMusical()).thenReturn(musical);
+
+    // 데이터베이스 초기화
```
마찬가지로 제거 부탁드려요

veniharuka replied:
반영하였습니다.

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceTest.java:
```diff
@@ -73,6 +106,62 @@ void setUp() {
+  }
+
+  private void initializeTestData() {
+    // 기존의 데이터베이스 초기화 로직 그대로 유지
```
마찬가지 입니다.

veniharuka replied:
반영하였습니다.

veniharuka committed:
Remove : Ticket 도메인 패키지의 exception java 파일 삭제

-GlobalExceptionHandler와 CustomException으로 통합으로 관리하기에 필요없는 파일 삭제

veniharuka committed:
Comment : 필요 없는 주석 삭제

- 의미 없는 주석 삭제

veniharuka committed:
Test: 티켓 서비스 동시성 테스트 가독성 및 일관성 개선

- initializeTestData를 setup으로 통합하여 테스트 데이터 초기화 일원화
- SQL 문법 소문자로 통일
- 불필요한 testPaymentSuccessButBookingFail 테스트 메서드 제거
- 시나리오에 맞게 testDoubleBookingAfterPayment 메서드 수정
- verifyDatabaseState, analyzeResults 메서드 적용

veniharuka committed:
Style: BookingResult의 userId 타입을 Long으로 변

- userId 필드 타입을 int에서 Long으로 변경
- userId 생성자 파라미터 및 getter 반환 타입도 함께 Long으로 변경

Why:
1. 테스트에서 사용하는 jdbcTemplate.queryForObject가 Long 타입을 반환
2. 데이터베이스의 id 컬럼이 BIGINT 타입으로 정의되어 있어 Long과 매핑
3. int(최대 약 21억)보다 큰 범위의 Long(약 922경)이 대용량 데이터 처리에 적합

Test:
- testDoubleBookingAfterPayment() 테스트에서 타입 불일치 오류 해결 확인

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
-        +
-        "(1, '서울', '서울 아트센터', 1, 150, '2024-10-16', '2024-10-01', '킹키부츠', null, null, null, null)");
+  private void initializeTestData() {
+    // 테스트용 회원 100명 생성
```
여기 틀렸다! 10명으로 바꿔주라 멤버 관련한거 전부

veniharuka replied:
해당 부분 수정하였습니다.

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -39,170 +47,355 @@ class TicketServiceConcurrencyTest {
+        "insert into section (id, available_seats, grade, price, musical_id, schedule_id) values "
+            + "(1, 50, 'R', 190000, 1, 1)");
 
-    // seat 데이터 삽입
```
좌석 1개 추가하는 부분 살리고 50개는 TicketServiceTest로 옮겨야할듯!

veniharuka replied:
for문 대신 기존의 좌석 1개 추가로 수정하였습니다.

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -39,170 +47,355 @@ class TicketServiceConcurrencyTest {
-      Long.class
-    );
-    assertEquals(2, memberCount, "Both members should be inserted successfully.");
+  @DisplayName("결제 완료 후 동시 티켓 할당으로 인한 중복 예매 발생 시나리오")
```
이거 리뷰 왜사라졌을까 이 테스트 지금
동일한 좌석에 여러 명이 예매하는 상황과 같아서 스레드 동시실행하는 부분만 여러명 예매하는 테스트에 통합시키고 이 시나리오는 삭제하는게 맞을듯!

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -39,170 +47,355 @@ class TicketServiceConcurrencyTest {
+    ConcurrentHashMap<String, BookingResult> results = new ConcurrentHashMap<>();
+    List<Long> bookingOrder = new ArrayList<>();
+
+// 첫 번째 사용자의 시도 시간과 결과
```
주석 불필요한거 다 지워주세요! 변수명과 중복돼서 가독성이 떨어지는 거 같습니다

veniharuka replied:
지웠습니다. 커밋 완료 후 확인 부탁드립니다.

veniharuka committed:
Test: testConcurrentSameSeatBooking 메서드 이동 및 설정 변경

testConcurrentSameSeatBooking 메서드를 TicketServiceConcurrencyTest.java로 이동
-초기 데이터 설정 개수 50에서 10으로 수정
-불필요한 주석 삭제

veniharuka committed:
Test : testConcurrentSameSeatBooking 메서드 이동 및 병합

TicketServiceTest에서 testConcurrentSameSeatBooking 메서드를 TicketServiceConcurrencyTest로 이동
- 기존 testBookTicket_singleTicketSuccess 메서드는 유지
- testDoubleBookingAfterPayment 메서드 삭제
- testDoubleBookingAfterPayment의 스레드 동시 실행 로직을 testConcurrentSameSeatBooking에 병합

chuseok commented on src/test/java/com/kb/wallet/ticket/service/TicketServiceConcurrencyTest.java:
```diff
@@ -1,208 +1,662 @@
+@ActiveProfiles("test")
 class TicketServiceConcurrencyTest {
 
+  private static final Logger log = LoggerFactory.getLogger(TicketServiceConcurrencyTest.class);
```
@Slf4j 사용 부탁해! build.gradle 설정해야될거야
설정 후에 global-setting으로 브랜치 생성해서 공유 부탁해

다른 부분도 수정해줘!

veniharuka committed:
Comment : 필요 없는 주석 삭제

- 의미 없는 주석 삭제

veniharuka committed:
Refactor: Logger 구현 변경 및 불필요한 주석 제거

- LoggerFactory를 Lombok @Slf4j로 교체
- 의미없는 주석 삭제

veniharuka committed:
Build: Lombok 테스트 환경 설정 추가

- Test scope에 Lombok 의존성 설정 추가
 - testCompileOnly
 - testAnnotationProcessor

veniharuka committed:
Merge branch 'main' into feature/ticketing/pre-concurrency/cmj
